### ISSUE TITLE
Consumer stop receive messages from broker

### ISSUE THREAD

**Issue by codelipenghui** [apache/pulsar#3131]
#### Actual behavior

A process subscribe a topic with 2 subscriptionName. After a few hours, we found backlog of 1 subscriptionName growing continuous, then check the broker stats:

![image](https://user-images.githubusercontent.com/12592133/49558266-6a859880-f945-11e8-9786-8f4cc9dc24c1.png)

![image](https://user-images.githubusercontent.com/12592133/49558323-9739b000-f945-11e8-8a6e-6f368e6bf7c3.png)

Stats of unackedMessages and availablePermits looks weird, i don't understand why both of them is 0 at the same time.

Then make the heap dump of client, but did not find anything abnormal. I'm already upload the heap dump to slack:

https://apache-pulsar.slack.com/files/UC0BYCXKQ/FELNFAZ6W/heap.hprof

#### System configuration

**Pulsar Broker version**: 2.2.0
**Pulsar Client version**: 2.2.0

**Comment by jiazhai** [comment]
Hi @codelipenghui, Do you think this is related with #3185?

**Comment by codelipenghui** [comment]
![image](https://user-images.githubusercontent.com/12592133/50676431-43e37c00-102f-11e9-9388-59792ec2ed57.png)
Consumer in the same subscription stop consume messages, seems like broker update available permits problem?

**Comment by codelipenghui** [comment]
![image](https://user-images.githubusercontent.com/12592133/50880859-cfd41a00-141a-11e9-84de-a466334c5c48.png)

Found some common problem cause consumer stop consume messages, stats of broker show us AVAILABLE PERMITS is negative number.

**Comment by codelipenghui** [comment]
This is a test case to show how consumer stop consume messages:

```java
package com.zhaopin.pulsar.issues;

import org.apache.pulsar.client.api.Consumer;
import org.apache.pulsar.client.api.Message;
import org.apache.pulsar.client.api.Producer;
import org.apache.pulsar.client.api.PulsarClient;
import org.apache.pulsar.client.api.PulsarClientException;
import org.apache.pulsar.client.api.SubscriptionType;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class ConsumerStopReceiveMessages {

    public static void main(String[] args) throws PulsarClientException {

        final String topic = "your-topic";

        /**
         * Broker configs:
         *
         * maxUnackedMessagesPerConsumer=500
         * maxUnackedMessagesPerSubscription=2000
         */
        PulsarClient client = PulsarClient.builder()
                .serviceUrl("pulsar://localhost:6650")
                .build();

        // Create a producer with enable message batching
        Producer<byte[]> producer = client.newProducer()
                .topic(topic)
                .batchingMaxMessages(1000)
                .batchingMaxPublishDelay(30, TimeUnit.SECONDS)
                .blockIfQueueFull(true)
                .maxPendingMessages(1000)
                .enableBatching(true)
                .create();

        // Create 6 consumers
        List<Consumer<byte[]>> consumers = new ArrayList<>();
        for (int i = 0; i < 6; i++) {
            consumers.add(client.newConsumer()
                    .topic(topic)
                    .subscriptionType(SubscriptionType.Shared)
                    .subscriptionName("test")
                    .ackTimeout(10, TimeUnit.SECONDS)
                    .receiverQueueSize(100)
                    .subscribe());
        }

        // Producer start publish messages
        new Thread(() -> {
            for (; ; ) {
                producer.sendAsync(("a message").getBytes());
            }
        }).start();

        // Consumers start consume messages
        consumers.forEach(consumer -> new Thread(() -> {
            do {
                // can't receive message
                try {
                    Message<byte[]> msg = consumer.receive();
                    System.out.println(consumer.getConsumerName() + " ---- " + new String(msg.getValue()));
                    Thread.sleep(50);
                    consumer.acknowledge(msg);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            } while (true);
        }).start());
    }
}

```

After a while, you will get following logs:

```
1cce6 ---- a message
1cce6 ---- a message
1cce6 ---- a message
1cce6 ---- a message
1cce6 ---- a message
1cce6 ---- a message
1cce6 ---- a message
[flink-source2] [test] [94eca] Prefetched messages: 0 --- Consume throughput: 0.53 msgs/s --- Throughput received: 0.00 msg/s --- 0.00 Mbit/s --- Ack sent rate: 0 ack/s --- Failed messages: 0 --- Failed acks: {}
1cce6 ---- a message
1cce6 ---- a message
[flink-source2] [test] [1cce6] Prefetched messages: 6812 --- Consume throughput: 6.37 msgs/s --- Throughput received: 0.00 msg/s --- 0.00 Mbit/s --- Ack sent rate: 0 ack/s --- Failed messages: 0 --- Failed acks: {}
1cce6 ---- a message
[ConsumerBase{subscription='test', consumerName='1cce6', topic='flink-source2'}] 7 messages have timed-out
[flink-source2] [test] [00c2e] Prefetched messages: 0 --- Consume throughput: 3.20 msgs/s --- Throughput received: 0.00 msg/s --- 0.00 Mbit/s --- Ack sent rate: 0 ack/s --- Failed messages: 0 --- Failed acks: {}
[flink-source2] [test] [1b009] Prefetched messages: 0 --- Consume throughput: 3.20 msgs/s --- Throughput received: 0.00 msg/s --- 0.00 Mbit/s --- Ack sent rate: 0 ack/s --- Failed messages: 0 --- Failed acks: {}
[flink-source2] [pulsar-api-test-14-3231] Pending messages: 1 --- Publish throughput: 7398.33 msg/s --- 0.51 Mbit/s --- Latency: med: 244.845 ms - 95pct: 588.543 ms - 99pct: 712.777 ms - 99.9pct: 851.404 ms - max: 851.478 ms --- Ack received rate: 7398.33 ack/s --- Failed messages: 0
[flink-source2] [test] [1cce6] Prefetched messages: 0 --- Consume throughput: 0.02 msgs/s --- Throughput received: 0.00 msg/s --- 0.00 Mbit/s --- Ack sent rate: 0 ack/s --- Failed messages: 0 --- Failed acks: {}
[flink-source2] [pulsar-api-test-14-3231] Pending messages: 1 --- Publish throughput: 6949.81 msg/s --- 0.48 Mbit/s --- Latency: med: 277.904 ms - 95pct: 609.112 ms - 99pct: 659.421 ms - 99.9pct: 746.124 ms - max: 746.239 ms --- Ack received rate: 6949.81 ack/s --- Failed messages: 0
[flink-source2] [pulsar-api-test-14-3231] Pending messages: 1 --- Publish throughput: 9166.40 msg/s --- 0.63 Mbit/s --- Latency: med: 163.221 ms - 95pct: 530.519 ms - 99pct: 651.178 ms - 99.9pct: 813.009 ms - max: 864.875 ms --- Ack received rate: 9166.40 ack/s --- Failed messages: 0

```

**Comment by jiazhai** [comment]
as discussed with @codelipenghui , above reproduce is under 2.2.0.

**Comment by bputt** [comment]
Any indication when this fix would get released, maybe 2.2.2?

**Comment by codelipenghui** [comment]
### Broker dispatch process:
1. Consumer subscribe to a topic, consumer will send a flow request.
2. By default, broker will read 100 entries from ledger and distribute entries to 5 consumers, one consumer will get 20 entries, in the test case, every entries has 2000 messages by enable the batching message in producer.
3. MaxUnackMessagesPerConsumer in broker settings is 100 and MaxUnackMessagesPerSubscription is 500.
4. So when broker send messages to the first consumer, dispatcher will be blocked, because 20 entries has 40000 messages. I have to consumers with receive queue size 100, so the dispatcher init available permits is 200, after send 40000 messages to the first consumer, available permits is 39980.
5. Consumer received messages and when call receive time > 50(receive queue size is 100), consumer will send permits request to the broker again, but now consumer is blocked by un-ack messages, so broker do not send any messages to consumer, and consumer will increase `permitsReceivedWhileConsumerBlocked`.
6. While consumer send a redelivery request, dispatcher should release the flow control, but the logic in the consumer, consumer will use min(`permitsReceivedWhileConsumerBlocked`,`totalRedelivery`) to increase the available permits in the dispatcher. I think the problem is here, if totalRedelivery < permitsReceivedWhileConsumerBlocked, consumer is already send flow request, but broker can't release the flow control. When a consumer receive 10000 messages(5 message, batch size is 2000), consumer will send 200 times flow request(50 messages per request),  if permitsReceivedWhileConsumerBlocked = 9000, at this time consumer redelivery 2000 messages, dispatcher will use 2000 to increase the available permits. after consumer send the last 1000 messages flow request, dispatcher is until blocked.

This is the test log:

[stop-consume-broker.log](https://github.com/apache/pulsar/files/2747607/stop-consume-broker.log)
